### PR TITLE
Cloud feeds

### PR DESCRIPTION
**Pull request by n-vr**
## What does this change

Added endpoints for cloud feeds.
- POST /cloud_feed - create a new cloud feed
- GET /account/{id}/cloud_feed_auth - get all available cloud feeds for an account, and if they are connected or not.
- POST /account/{id}/cloud_feed_auth - send an auth grant token

Changed endpoints:
- POST /campaign - added cloud_feeds field

## Testing

- [ ] This change does not need testing.
- [x] I tested this change.
- [ ] I did not test this change.

I started a local development environment and made the calls.

## Trello card(s)

- Implement OAuth2 Authorisation grant (at least for Enelogic): https://trello.com/c/2giM5mkF

### CODE REVIEW

File: repositories/app.go [hunk 1]
```diff
@@ -20,6 +20,7 @@ type AppModel struct {
 	gorm.Model
 	Name                    string `gorm:"unique;not null"`
 	ProvisioningURLTemplate string
+	OauthRedirectURL        string
```
Comment (rickklaasboer) [inline repositories/app.go:23]:
```suggestion
	OAuthRedirectURL        string
```

File: repositories/app.go [hunk 2]
```diff
@@ -33,6 +34,7 @@ func MakeAppModel(app twomes.App) AppModel {
 		Model:                   gorm.Model{ID: app.ID},
 		Name:                    app.Name,
 		ProvisioningURLTemplate: app.ProvisioningURLTemplate,
+		OauthRedirectURL:        app.OauthRedirectURL,
```
Comment (rickklaasboer) [inline repositories/app.go:37]:
```suggestion
		OAuthRedirectURL:        app.OauthRedirectURL,
```

File: repositories/app.go [hunk 3]
```diff
@@ -42,6 +44,7 @@ func (m *AppModel) fromModel() twomes.App {
 		ID:                      m.Model.ID,
 		Name:                    m.Name,
 		ProvisioningURLTemplate: m.ProvisioningURLTemplate,
+		OauthRedirectURL:        m.OauthRedirectURL,
```
Comment (rickklaasboer) [inline repositories/app.go:47]:
```suggestion
		OAuthRedirectURL:        m.OauthRedirectURL,
```

File: repositories/cloudfeedauth.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+// Database representation of [twomes.CloudFeedAuth].
+type CloudFeedAuthModel struct {
+	AccountID   uint `gorm:"primaryKey;autoIncrement:false"`
+	CloudFeedID uint `gorm:"primaryKey;autoIncrement:false"`
+	CreatedAt   time.Time
+	UpdatedAt   time.Time
+	DeletedAt   gorm.DeletedAt `gorm:"index"`
+	// TODO: WARNING encrypted sting encryption not yet implemented.
```
Comment (rickklaasboer) [inline repositories/cloudfeedauth.go:28]:
```suggestion
	// TODO: WARNING encrypted string encryption not yet implemented.
```

File: services/app.go [hunk 5]
```diff
@@ -17,8 +17,8 @@ func NewAppService(repository ports.AppRepository) *AppService {
 }
 
 // Create a new app.
-func (s *AppService) Create(name string, provisioningURLTemplate string) (twomes.App, error) {
-	app := twomes.MakeApp(name, provisioningURLTemplate)
+func (s *AppService) Create(name, provisioningURLTemplate, oauthRedirectURL string) (twomes.App, error) {
+	app := twomes.MakeApp(name, provisioningURLTemplate, oauthRedirectURL)
```
Comment (rickklaasboer) [inline services/app.go:21]:
```suggestion
func (s *AppService) Create(name, provisioningURLTemplate, oAuthRedirectURL string) (twomes.App, error) {
	app := twomes.MakeApp(name, provisioningURLTemplate, oAuthRedirectURL)
```

File: twomes/app.go [hunk 6]
```diff
@@ -5,12 +5,14 @@ type App struct {
...
+	OauthRedirectURL        string `json:"oauth_redirect_url"`
 }
 
 // Create a new app.
-func MakeApp(name, provisioningURLTemplate string) App {
+func MakeApp(name, provisioningURLTemplate, oauthRedirectURL string) App {
 	return App{
 		Name:                    name,
 		ProvisioningURLTemplate: provisioningURLTemplate,
+		OauthRedirectURL:        oauthRedirectURL,
```
Comment (rickklaasboer) [inline twomes/app.go:16]:
```suggestion
func MakeApp(name, provisioningURLTemplate, oAuthRedirectURL string) App {
	return App{
		Name:                    name,
		ProvisioningURLTemplate: provisioningURLTemplate,
		OAuthRedirectURL:        oAuthRedirectURL,
```

File: twomes/app.go [hunk 7]
```diff
@@ -5,12 +5,14 @@ type App struct {
 	ID                      uint   `json:"id"`
 	Name                    string `json:"name"`
 	ProvisioningURLTemplate string `json:"provisioning_url_template"`
+	OauthRedirectURL        string `json:"oauth_redirect_url"`
```
Comment (rickklaasboer) [inline twomes/app.go:8]:
```suggestion
	OAuthRedirectURL        string `json:"oauth_redirect_url"`
```

**Review by rickklaasboer — COMMENTED** [review]
Found some inconsistencies in casing.

**Comment by n-vr** [comment]
> A cloud feed needs to be linked to a campaign for it to be seen returned in the `/account/<id>/cloud_feed_auth` endpoint, right? How do I manage this using the API?

See the OpenAPI spec for POST /campaign endpoint.
These endpoints do not have a PUT method implemented yet, so at this time, if you already created a campaign, you need to add a new line in the database  in the `campaign_cloud_feed` table.

**Pull request merged**